verification/io/io-spec.gobra

```diff
@@ -38,75 +38,191 @@ type Place int
-if asid ∈ core ∧ (∃currseg nextseg.
-ConsDir nextseg = ConsDir currseg ∧ (∃traversedseg intermediatepkt hf1 hf2.
-xover_core2_link_type hf1 hf2 asid (ConsDir currseg) ∧
-(∃nextfut. dp3s_xover_common s pkt currseg nextseg traversedseg intermediatepkt hf1 hf2 nextfut recvif newpkt nextif)))
-then ExT (λT'. ExV (λX. CBio (IN bio3s_xover_core) t (Internal_val1 pkt recvif newpkt nextif) X
-T' * dp3s_iospec_ordered (dp3s_add_obuf s nextif newpkt) T')) else Bool True
-| _ ⇒ Bool True)
-*/
-
-pred dp3s_iospec_bio3s_exit(s IO_dp3s_state_local, t Place)
-/* s t ≡ ∀⋆ (λv. case v of
-Internal_val2 pkt newpkt nextif ⇒
-if pkt ∈ ibuf s None ∧ dp3s_forward_ext pkt newpkt nextif
-then ExT (λT'. ExV (λX. CBio (IN bio3s_exit) t (Internal_val2 pkt newpkt nextif) X T' ⋆ dp3s_iospec_ordered (dp3s_add_obuf s (Some nextif) newpkt) T'))
-else Bool True | _ ⇒ Bool True)
-*/
-
-pred CBioIO_bio3s_send(t Place)
+pred CBio_IN_bio3s_enter(t Place, v IO_val)
 
 ghost
-requires CBioIO_bio3s_send(t)
-pure func dp3s_iospec_bio3s_send_T(s IO_dp3s_state_local, t Place) Place
+requires CBio_IN_bio3s_enter(t, v)
+decreases
+pure func CBio_IN_bio3s_enter_T(s IO_dp3s_state_local, t Place, v IO_val) Place
 
+
+// This corresponds to the condition of the if statement in the io-spec case for enter
+ghost
+decreases
+pure func dp3s_iospec_bio3s_enter_guard(s IO_dp3s_state_local, t Place, v IO_val) bool {
+	return some(v.IO_Internal_val1_2) in domain(s.ibuf) ==>
+		(let ibuf_set := s.ibuf[some(v.IO_Internal_val1_2)] in (v.IO_Internal_val1_1 in ibuf_set)) &&
+		(exists currseg IO_seg3, traversedseg IO_seg3 ::
+		(exists hf1 IO_HF, fut seq[IO_HF] ::
+			dp2_enter_guard(
+				v.IO_Internal_val1_1,
+				currseg,
+				traversedseg,
+				asid(),
+				hf1,
+				v.IO_Internal_val1_2,
+				fut) &&
+				dp3s_forward(
+					IO_pkt2(
+						IO_Packet2{
+							traversedseg,
+							v.IO_Internal_val1_1.LeftSeg,
+							v.IO_Internal_val1_1.MidSeg,
+							v.IO_Internal_val1_1.RightSeg}),
+					v.IO_Internal_val1_3,
+					v.IO_Internal_val1_4)))
+}
+
+pred dp3s_iospec_bio3s_enter(s IO_dp3s_state_local, t Place) {
+	forall v IO_val ::
+	(v.isIO_Internal_val1 ==>
+		(dp3s_iospec_bio3s_enter_guard(s, t, v) ==>
+			(CBio_IN_bio3s_enter(t, v) &&
+				dp3s_iospec_ordered(
+					dp3s_add_obuf(s, v.IO_Internal_val1_4, v.IO_Internal_val1_3),
+					CBio_IN_bio3s_enter_T(s, t, v)))))
+}
+
+pred CBio_IN_bio3s_xover_up2down(t Place, v IO_val)
+
+ghost
+requires CBio_IN_bio3s_xover_up2down(t, v)
+pure func dp3s_iospec_bio3s_xover_up2down_T(s IO_dp3s_state_local, t Place, v IO_val) Place
+
+
+// This corresponds to the condition of the if statement in the io-spec case for xover_up2down
+ghost
+decreases
+pure func dp3s_iospec_bio3s_xover_up2down_guard(s IO_dp3s_state_local, t Place, v IO_val) bool {
+	return (exists currseg IO_seg3 :: !currseg.ConsDir &&
+		(exists nextseg IO_seg3 :: nextseg.ConsDir &&
+			(exists traversedseg IO_seg3, intermediatepkt IO_pkt3, hf1 IO_HF, hf2 IO_HF ::
```

Comment:
If we follow the definition of `xover_up2down2_link_type`, we will also find that most of the variables that are existentially quantified (if not all) manifest a functional dependency on values that are universally quantified, so we can in principle get rid of these existentials as well